src/components/abstractComponent.ts

```diff
@@ -0,0 +1,436 @@
+
+  public render() {
+    if (this._isAnchored && this._isSetup && this.width() >= 0 && this.height() >= 0) {
+      RenderController.registerToRender(this);
+    }
+    return this;
+  }
+
+  public onResize(resizeHandler: IResizeHandler) {
+    this._resizeHandler = resizeHandler;
+    return this;
+  }
+
+  abstract renderImmediately(): this;
+
+  abstract redraw(): this;
+
+  abstract renderTo(container: D): this;
+
+  public xAlignment(): string;
+  public xAlignment(xAlignment: string): this;
+  public xAlignment(xAlignment?: string): this | string {
+    if (xAlignment == null) {
+      return this._xAlignment;
+    }
+
+    xAlignment = xAlignment.toLowerCase();
+    if (AbstractComponent._xAlignToProportion[xAlignment] == null) {
+      throw new Error("Unsupported alignment: " + xAlignment);
+    }
+    this._xAlignment = xAlignment;
+    this.redraw();
+    return this;
+  }
+
+  public yAlignment(): string;
+  public yAlignment(yAlignment: string): this;
+  public yAlignment(yAlignment?: string): this | string {
+    if (yAlignment == null) {
+      return this._yAlignment;
+    }
+
+    yAlignment = yAlignment.toLowerCase();
+    if (AbstractComponent._yAlignToProportion[yAlignment] == null) {
+      throw new Error("Unsupported alignment: " + yAlignment);
+    }
+    this._yAlignment = yAlignment;
+    this.redraw();
+    return this;
+  }
+
+  public hasClass(cssClass: string) {
+    if (cssClass == null) {
+      return false;
+    }
+
+    if (this._element == null) {
+      return this._cssClasses.has(cssClass);
+    } else {
+      return this._element.classed(cssClass);
+    }
+  }
+
+  public addClass(cssClass: string) {
+    if (cssClass == null) {
+      return this;
+    }
+
+    if (this._element == null) {
+      this._cssClasses.add(cssClass);
+    } else {
+      this._element.classed(cssClass, true);
+    }
+
+    return this;
+  }
+
+  public removeClass(cssClass: string) {
+    if (cssClass == null) {
+      return this;
+    }
+
+    if (this._element == null) {
+      this._cssClasses.delete(cssClass);
+    } else {
+      this._element.classed(cssClass, false);
+    }
+
+    return this;
+  }
+
+  /**
+   * Checks if the Component has a fixed width or if it grows to fill available space.
+   * @default false
+   */
+  public fixedWidth() {
+    return false;
+  }
+
+  /**
+   * Checks if the Component has a fixed height or if it grows to fill available space.
+   * @default false
+   */
+  public fixedHeight() {
+    return false;
+  }
+
+  abstract detach(): this;
+
+  public onDetach(callback: GenericComponentCallback<D>) {
+    this._onDetachCallbacks.add(callback);
+    return this;
+  }
+
+  public offDetach(callback: GenericComponentCallback<D>) {
+    this._onDetachCallbacks.delete(callback);
+    return this;
+  }
+
+  public parent(): ComponentContainer;
+  public parent(parent: ComponentContainer): this;
```

Comment:
Since `ComponentContainer` is an HTMLComponent, does this mean we can't nest SVG components inside other SVG components?